substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/debugentry/DebugInfoBase.java

```diff
@@ -419,19 +423,24 @@ private Range addSubrange(DebugLocationInfo locationInfo, Range primaryRange, Cl
     }
 
     private ClassEntry ensureClassEntry(ResolvedJavaType type) {
-        /* See if we already have an entry. */
-        ClassEntry classEntry = primaryClassesIndex.get(type);
+        /* We should already have an entry -- TODO prove that claim. */
+        ClassEntry classEntry = instanceClassesIndex.get(type);
         if (classEntry == null) {
+            /* We must have a type entry -- TODO prove we never reach here. */
```

Comment:
I am confident we can now reduce this call to a lookup, with an assert that the result is found, and dispense with any need for proof. So, I'll push an extra change to that effect.

Why so? Well, the continued existence of this method is an artefact of creeping change in the meaning of the fields now renamed to `instancesClasses` and `instanceClassesIndex`. Originally these fields tracked instance classes with top level compiled methods. The methods and associated classes were presented via the DebugCodeInfo API and ultimately derived by iterating over the `NativeImageCodeCache`. The expectation was that all of these classes would already have been presented via the DebugTypeInfo API, ultimately derived by iterating over the Universe type. However, the list was originally only meant to include the proper (i.e. strict) subset of the notified instance classes which needed to be qualified with method location and range info.

When we started reporting frame data via the `DebugCodeInfo` API we found classes that were i) not already identified via the `DebugTypeInfo` API or ii) had inlined methods but no top level methods. These also required generation of inline method location and range info. So, we threw them into the list and index, creating a `ClassEntry` if none existed and then sorted them out case by case in code that used the list of classes.

The first problem was an input hygiene problem. We ran into 'missing' types because frames were actually referencing analysis types instead of hosted types. That has been fixed on the notification side (`NativeImageDebugInfoProvider`) by remapping classes from analysis to hosted domain before passing them through the `DebugCodeInfo` API. 

The second problem reflect the fact that we really need to iterate over all instance classes at certain points in the code and only over some instance classes at others. A `ClassEntry` now identifies whether or not it has top level compiled methods and/or inlined methods and all code that now consumes the `instanceClasses` list tests for the relevant cases.

So, both the problems we originally faced appear to be resolved. The current code populates this list and index with all expected instance classes during traversal of universe types set. So, the checks and updtes that used to be done in `ensureClassEntry` really ought to be redundant.

n.b. There is another follow-up task needed to consolidate the current position beyond redefining `ensureClassEntry`. From one point of view fields `instanceClasses` and `instanceClassesIndex` are redundant. We now only consume the instaceClasses list via a Stream API. The stream content manifested as a list in `instanceClasses` can be derived virtually by filtering field `types` with a type test for `ClassEntry` and cast before returning. Similarly, lookups in the index map `instanceClassesIndex` can be replaced with a lookup in the index map `typesIndex` that also employs an `instanceof` check. So, we have the option to virtualize the instance class list and index, trading off space against a small amount of stream processing time.

Whichever way we decide to go with this a code change is needed. There are a couple of cases left where iteration over all instance classes was and still is performed by streaming and filtering field `types` (this made sense when not all instance classes were listed in field `instanceClasses`).  I'm happy to fix this as part of this patch or in a follow-up. Do you have a preference for retaining two lists or virtualizing one of them?

Reply:
Fixed. I retained the separate list of instance classes but everything is now using streams derived either from the types list or from the classes list.